### ISSUE TITLE
Ambiguities? Please explain or fix.

### ISSUE THREAD

**Issue by RieksJ** [AmpersandTarski/Ampersand#246]
Consider the following script:

```
 CONTEXT PleaseExplainError IN ENGLISH

 r :: SESSION * A
 s :: SESSION * B

 RULE "ambiguities?": '_SESSION';s |- -r;V

 ENDCONTEXT
```

This script produces the following output 

```
 Error(s) found:
 An ambiguity arises in type checking. Be more specific in the expressions -r and V[A*B].
   You could add more types inside the expression, or write:
   (V[SESSION*A] - r)
   (V[A*B]) /\ V[A*B]
   Error at symbol () in file ticket.adl at line 6 : 40
```

I do not understand why there is an ambiguity here. What other options are there more than what I consider the trivial one: `(V[SESSION*A]-r);V[A*B]`?

B.t.w., I use Ampersand v3.3.0[master:1e93f51], which is based on the new Haskell stuff.

**Comment by sjcjoosten** [comment]
The issue is that V is not bound in the expression, and semantics may change because of it. This is not allowed since https://github.com/AmpersandTarski/ampersand/pull/241 and I agree that the error message should be improved.

**Comment by RieksJ** [comment]
Huh? I'm a big confused. My first reading of your comment was that BECAUSE you and #241 (which is you) are in agreement (which is nice/consistent), something is not allowed, but to me there is this ambiguity: do your refer to `V` not having its signature specified, or there a missing `V` is before the `-` operator? Or is this your hypnotic suggestion to say that you are going to improve the error message? `:-)`

**Comment by WolframKahl** [comment]
On Mon, Dec 14, 2015 at 12:07:10AM -0800, Sebastiaan Joosten wrote:
> The issue is that V is not bound in the expression,

What does ``bound'' mean here? I am used to ``bound variables'',
but since V is a (parameterised) constant, this sentence sounds
extremely strange to me.


Wolfram

**Comment by sjcjoosten** [comment]
Sorry, let me rephrase.

The issue is that in (V[A*C]-r);V[C*B] is a correct typing, provided that C ISA B. If there are multiple choice, the type checker will usually pick a most obvious one (usually the smallest, which would be B). However in this, the choice of C determines the outcome of the expression, so we throw an error.

A problem with the error message is that it is delayed until it actually occurs. Expressions like (-r);s would be fine, and are accepted, since any choice A, B of (V[A*B]-r) will do, as long as V[A*B] contains r. Consequently, the place where the error becomes visible is not the part of the expression the error should be about.

I don't know what to call this exactly. The word bound (Dutch: begrenst) was intended to mean bound in the sense of upper-bound, not in the sense of bound variable. I realise that is a very unfortunate choice of words. Any suggestions for something that indicates whether the expression might change, given a more general type?

It will take some time (around a month) before I am able to get to this (in terms of improving the message).

**Comment by sjcjoosten** [comment]
Ah google translate is my friend. 'Confined' seems to be the word to use here (other suggestions include restricted and limited)

**Comment by RieksJ** [comment]
@sjcjoosten You write "The issue is that in (V[AC]-r);V[CB] is a correct typing, provided that C ISA B". Why do we need "provided that C ISA B"? What's wrong when C is not a B?

**Comment by WolframKahl** [comment]
On Mon, Dec 14, 2015 at 06:12:50AM -0800, Sebastiaan Joosten wrote:
 > The issue is that in (V[A*C]-r);V[C*B] is a correct typing, provided
 > that C ISA B. If there are multiple choice, the type checker will
 > usually pick a most obvious one (usually the smallest, which would be
 > B). However in this, the choice of C determines the outcome of the
 > expression, so we throw an error.
 > 
 > A problem with the error message is that it is delayed until it
 > actually occurs. Expressions like (-r);s would be fine, and are
 > accepted, since any choice A, B of (V[A*B]-r) will do, as long as
 > V[A*B] contains r. Consequently, the place where the error becomes
 > visible is not the part of the expression the error should be about.

Thank you for the clarification!

 > I don't know what to call this exactly. The word bound (Dutch:
 > begrenst) was intended to mean bound in the sense of upper-bound

I think that ``bounded'' might be an appropriate translation.
Then ``boundness'' in the referenced pull would become
``boundedness'', and also make more sense.
I am still not sure whether this is an optimal terminology,
but at least I find it acceptable.

 > I realise that is a very unfortunate
 > choice of words. Any suggestions for something that indicates whether
 > the expression might change, given a more general type?

That is a more general concept --- I'll think about it.


Wolfram

**Comment by sjcjoosten** [comment]
@RieksJ I believe we check compatibility of the type of ``a`` and ``b` in ``a - b``

**Comment by RieksJ** [comment]
From phone contacts with @sjcjoosten and @stefjoosten, I hear two arguments for generating the ambiguity error:

* ***if*** there were a `C` with `C ISA B`, then `(V[A*C]-r);V[C*B]` would be a correct typing, and the original statement `-r;V` would be ambiguous; and since there is no way of knowing whether or not there will ever be such a `C`, we choose to generate the error so that if `C` comes alive, we won't have to go back to disambiguate `-r;V`.

I prefer to have errors informing me about all the places where it 'hurts' if at some point in the future `C` were to come into existence (so I can fix them) rather than being forced up-front to fix 'errors' that may (or may not) occur some time in the future. If the argument of imposing constraints to prevent future errors were generallly valid, then Ampersand would become useless as it is impossible to write a (non-empty or empty) script that cannot be modified to contain an error.

* Currently, there are 'hidden' concepts, which come with the `REPRESENT` statement (e.g. 'Object', 'ALPHANUMERIC'). Since representations are done as a generalization, every concept that is defined in a script has a larger concept, either being 'Object' or some primitive type. The answer to the original question: "why is `-r;V` ambiguous?" would then be: Because it could mean `V[SESSION*A]-r;V[A*B]`, as well as `V[SESSION*Object]-r;V[Object*B]`.

I value this argument, and I remember there is some ticket(s) that made this happen. However, I would then like to see that this is made official, meaning that 'Object's and primitive concepts are treated as real concepts (as it used to be if I remember correctly). This would mean that this stuff gets a place in FormalAmpersand. It might mean that the syntax `REPRESENT {Concept} TYPE {PrimitiveConcept}` is replaced by `CLASSIFY {Concept} ISA {PrimitiveConcept}`, leaving the contents of the generics-files unchanged.

**Comment by stefjoosten** [comment]
Rieks, ik meen me te herinneren dat er iets bestaat om de Exec-Engine een tijdstempel te laten zetten. Heb je daar een voorbeeld van? (In het GitBook is het niet gedocumenteerd, dus voor hetzelfde geld bestaat het niet...)

Disclaimer
Dit bericht met eventuele bijlagen is vertrouwelijk en uitsluitend bestemd voor de geadresseerde. Indien u niet de bedoelde ontvanger bent, wordt u verzocht de afzender te waarschuwen en dit bericht met eventuele bijlagen direct te verwijderen en/of te vernietigen. Het is niet toegestaan dit bericht en eventuele bijlagen te vermenigvuldigen, door te sturen, openbaar te maken, op te slaan of op andere wijze te gebruiken. Ordina N.V. en/of haar groepsmaatschappijen accepteren geen verantwoordelijkheid of aansprakelijkheid voor schade die voortvloeit uit de inhoud en/of de verzending van dit bericht.

This e-mail and any attachments are confidential and are solely intended for the addressee. If you are not the intended recipient, please notify the sender and delete and/or destroy this message and any attachments immediately. It is prohibited to copy, to distribute, to disclose or to use this e-mail and any attachments in any other way. Ordina N.V. and/or its group companies do not accept any responsibility nor liability for any damage resulting from the content of and/or the transmission of this message.

**Comment by sjcjoosten** [comment]
@RieksJ can we close this ticket? It seems that we are now happy with the fact that there is a message, and the issues with FormalAmpersand have been moved to another ticket.

**Comment by RieksJ** [comment]
@sjcjoosten You agreed that the error message should be improved, not just 'be there' as in the situation that started the ticket. I suggest we close this ticket when the improved error message  is in place. Whatsay?

**Comment by sjcjoosten** [comment]
Any suggestions for the message?

**Comment by RieksJ** [comment]
That's a difficult one. Not really now.

However, some enhancements could be made:
- it would be helpful to not just refer to 'the expression' (to be possibly replaced), but actually mention that expression explicitly, e.g. as in `replace -r;V  with (V[SESSION*A] - r)`.
- suggestions such as `(V[A*B]) /\ V[A*B]` are not useful/helpful, so should not be shown.
- if you show expressions such as `(V[A*B]) /\ V[A*B]`, make sure parentheses are matched.

**Comment by sjcjoosten** [comment]
How about this:

``Error(s) found:
 An ambiguity arises in type checking -r;V
 Be more specific in the expressions -r and V.
   You could add more types inside the expression, for instance:
   (V[SESSION*A] - r);V
   -r;(V /\ V[A*B])``

**Comment by RieksJ** [comment]
How about this:
~~~
Error(s) found:
 An ambiguity arises when type checking the expression '-r;V'
   To resolve this ambiguity you may consider replacing this expression, e.g. by:
   (V[SESSION*A] - r);V
   -r;V[A*B]
~~~
